In [1]:
import csv
import numpy
from sklearn import tree
from sklearn import linear_model
from sklearn import svm

counter = 0
group = 0
sitting_values = {}
running_values = {}
stairs_values = {}
car_values = {}
walking_values = {}

# Read csv files for each kind of data and pull the necessary rows.
for line in csv.reader(open("sitting.csv")):
	if group < 30 and "Time since start in ms" not in line[6]:
		if counter < 100:
			counter = counter + 1
			if group in sitting_values:
				sitting_values[group].append(float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip())))
			else:
				sitting_values[group] = [float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip()))]
		else: 
			group = group + 1
			counter = 0

counter = 0
group = 0
for line in csv.reader(open("running.csv")):
	if group < 30 and "Time since start in ms" not in line[6]:
		if counter < 100:
			counter = counter + 1
			if group in running_values:
				running_values[group].append(float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip())))
			else:
				running_values[group] = [float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip()))]
		else: 
			group = group + 1
			counter = 0

counter = 0
group = 0
for line in csv.reader(open("walking.csv")):
	if group < 30 and "Time since start in ms" not in line[6]:
		if counter < 100:
			counter = counter + 1
			if group in walking_values:
				walking_values[group].append(float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip())))
			else:
				walking_values[group] = [float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip()))]
		else: 
			group = group + 1
			counter = 0

counter = 0
group = 0
for line in csv.reader(open("car.csv")):
	if group < 30 and "Time since start in ms" not in line[6]:
		if counter < 100:
			counter = counter + 1
			if group in car_values:
				car_values[group].append(float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip())))
			else:
				car_values[group] = [float(line[2].strip())+float(line[1].strip())+float(line[0].strip())]
		else: 
			group = group + 1
			counter = 0

counter = 0
group = 0
for line in csv.reader(open("stairs.csv")):
	if group < 30 and "Time since start in ms" not in line[6]:
		if counter < 100:
			counter = counter + 1
			if group in stairs_values:
				stairs_values[group].append(float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip())))
			else:
				stairs_values[group] = [float(line[2].strip())+float(line[1].strip())+abs(float(line[0].strip()))]
		else: 
			group = group + 1
			counter = 0

sitting_averages = {}
running_averages = {}
walking_averages = {}
car_averages = {}
stairs_averages = {}

sitting_stdevs = {}
running_stdevs = {}
walking_stdevs = {}
car_stdevs = {}
stairs_stdevs = {}

# Find the average and the standard deviations for each 10 second interval. 
for group in sitting_values:
	sitting_averages[group] = numpy.mean(sitting_values[group])
	sitting_stdevs[group] = numpy.std(sitting_values[group])
	
	running_stdevs[group] = numpy.std(running_values[group])
	running_averages[group] = numpy.mean(running_values[group])

	walking_stdevs[group] = numpy.std(walking_values[group])
	walking_averages[group] = numpy.mean(walking_values[group])

	car_stdevs[group] = numpy.std(car_values[group])
	car_averages[group] = numpy.mean(car_values[group])

	stairs_stdevs[group] = numpy.std(stairs_values[group])
	stairs_averages[group] = numpy.mean(stairs_values[group])

# create the training labels list for multiclass classifiers
training_labels = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4]

# create the training labels list for binary one vs rest classifiers
training_labels_categories =[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]

training_sitting = []
training_running = []
training_stairs = []
training_walking = []
training_car = []

# put the training values in the correct format for the classifiers
for i in range(0,24):
	training_sitting.append([1, sitting_averages[i]])
	training_running.append([1, running_averages[i]])
	training_stairs.append([1, stairs_averages[i]])
	training_walking.append([1, walking_averages[i]])
	training_car.append([1, car_averages[i]])

test_sitting = []
test_running = []
test_stairs = []
test_walking = []
test_car = []

# put the test values in the correct format for the classifiers
for i in range(24, 30):
	test_running.append([1, running_averages[i]])
	test_sitting.append([1, sitting_averages[i]])
	test_stairs.append([1, stairs_averages[i]])
	test_walking.append([1, walking_averages[i]])
	test_car.append([1, car_averages[i]])


# build the list of test features for multiclass classifier
all_test = test_sitting + test_walking+ test_running + test_car + test_stairs

# build the list of test features for the binary one vs rest classifiers
test_set_sitting = test_sitting + test_walking+ test_running + test_car + test_stairs
test_set_running = test_running +test_sitting + test_walking+ test_car + test_stairs
test_set_walking = test_walking+ test_running +test_sitting + test_car + test_stairs
test_set_stairs = test_stairs+test_walking+ test_running +test_sitting + test_car 
test_set_car = test_car +test_stairs+test_walking+ test_running +test_sitting 

# build the list of training features for multiclass classifier
all_training = training_sitting + training_walking + training_running + training_car  + training_stairs

# build the list of training features for the binary one vs rest classifiers
training_set_sitting = training_sitting + training_walking + training_running + training_car  + training_stairs
training_set_running = training_running + training_car  + training_stairs+training_sitting + training_walking
training_set_walking = training_walking + training_running + training_car  + training_stairs+training_sitting
training_set_stairs = training_stairs+training_sitting + training_walking + training_running + training_car  
training_set_car =  training_car+training_stairs+training_sitting + training_walking + training_running   

test_labels = [0,0,0,0,0,0,1,1,1,1,1,1,2,2,2,2,2,2,3,3,3,3,3,3,4,4,4,4,4,4]
test_label_categories = [0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]


# ======= Decision Tree Classifier =======
# fit the classifier
clf = tree.DecisionTreeClassifier()
clf = clf.fit(all_training, training_labels)
output_labels = []
for i in range(len(test_labels)):
	result = clf.predict([all_test[i]])
	output_labels.append(result[0])

# calculate the accuracy scores and print them
print "======= Decision Tree Classifier ======="
print ("Accuracy for multiclass is %.2f" % clf.score(all_test, test_labels))
print 'Predicted labels my MultiClass are: \t', output_labels
print 'Actual test labels should be: \t\t', test_labels
print 'Where 0 is sitting, 1 is walking, 2 is running, 3 is car, 4 is stairs.'
print
clf_sitting = tree.DecisionTreeClassifier()
clf_sitting = clf_sitting.fit(training_set_sitting, training_labels_categories)

clf_running = tree.DecisionTreeClassifier()
clf_running = clf_running.fit(training_set_running, training_labels_categories)

clf_stairs = tree.DecisionTreeClassifier()
clf_stairs = clf_stairs.fit(training_set_stairs, training_labels_categories)

clf_car= tree.DecisionTreeClassifier()
clf_car = clf_car.fit(training_set_car, training_labels_categories)

clf_walking = tree.DecisionTreeClassifier()
clf_walking = clf_walking.fit(training_set_walking, training_labels_categories)

print ("One vs Rest sitting accuracy is %.2f" % clf_sitting.score(test_set_sitting, test_label_categories))
print ("One vs Rest walking accuracy is %.2f" % clf_walking.score(test_set_walking, test_label_categories))
print ("One vs Rest running accuracy is %.2f" % clf_running.score(test_set_running, test_label_categories))
print ("One vs Rest car accuracy is %.2f" % clf_car.score(test_set_car, test_label_categories))
print ("One vs Rest stairs accuracy is %.2f" % clf_stairs.score(test_set_stairs, test_label_categories))


# ======= SVM Classifier =======
# fit the classifier
svc = svm.SVC(kernel='linear')
svc.fit(all_training, training_labels)  
output_labels = []
for i in range(len(test_labels)):
	result = svc.predict([all_test[i]])
	output_labels.append(result[0])

# calculate the accuracy scores and print them
print
print "======= SVM Classifier ======="
print ("Accuracy for MultiClass is %.2f" % svc.score(all_test, test_labels))
print  'Predicted labels my MultiClass are: \t', output_labels
print 'Actual test labels should be: \t\t', test_labels
print 'Where 0 is sitting, 1 is walking, 2 is running, 3 is car, 4 is stairs.'
print
svc_sitting = svm.SVC(kernel='linear')
svc_sitting = svc_sitting.fit(training_set_sitting, training_labels_categories)

svc_running = svm.SVC(kernel='linear')
svc_running = svc_running.fit(training_set_running, training_labels_categories)

svc_stairs = svm.SVC(kernel='linear')
svc_stairs = svc_stairs.fit(training_set_stairs, training_labels_categories)

svc_car= svm.SVC(kernel='linear')
svc_car = svc_car.fit(training_set_car, training_labels_categories)

svc_walking = svm.SVC(kernel='linear')
svc_walking = svc_walking.fit(training_set_walking, training_labels_categories)

print ("One vs Rest sitting accuracy is %.2f" % svc_sitting.score(test_set_sitting, test_label_categories))
print ("One vs Rest walking accuracy is %.2f" % svc_walking.score(test_set_walking, test_label_categories))
print ("One vs Rest running accuracy is %.2f" % svc_running.score(test_set_running, test_label_categories))
print ("One vs Rest car accuracy is %.2f" % svc_car.score(test_set_car, test_label_categories))
print ("One vs Rest stairs accuracy is %.2f" % svc_stairs.score(test_set_stairs, test_label_categories))
print


# ======= LogisticRegression Classifier =======
# fit the classifier
logistic = linear_model.LogisticRegression(C=1e5)
logistic.fit(all_training, training_labels)
output_labels = []
for i in range(len(test_labels)):
	result = logistic.predict([all_test[i]])
	output_labels.append(result[0])

# calculate the accuracy scores and print them

print "======= LogisticRegression Classifier ======="
print ("Accuracy for MultiClass is %.2f" %logistic.score(all_test, test_labels))
print 'Predicted labels my MultiClass are: \t', output_labels
print 'Actual test labels should be: \t\t', test_labels
print 'Where 0 is sitting, 1 is walking, 2 is running, 3 is car, 4 is stairs.'
print
logistic_sitting = linear_model.LogisticRegression(C=1e5)
logistic_sitting = logistic_sitting.fit(training_set_sitting, training_labels_categories)

logistic_running = linear_model.LogisticRegression(C=1e5)
logistic_running = logistic_running.fit(training_set_running, training_labels_categories)

logistic_stairs = linear_model.LogisticRegression(C=1e5)
logistic_stairs = logistic_stairs.fit(training_set_stairs, training_labels_categories)

logistic_car= linear_model.LogisticRegression(C=1e5)
logistic_car = logistic_car.fit(training_set_car, training_labels_categories)

logistic_walking = linear_model.LogisticRegression(C=1e5)
logistic_walking = logistic_walking.fit(training_set_walking, training_labels_categories)

print ("One vs Rest sitting accuracy is %.2f" % logistic_sitting.score(test_set_sitting, test_label_categories))
print ("One vs Rest walking accuracy is %.2f" % logistic_walking.score(test_set_walking, test_label_categories))
print ("One vs Rest running accuracy is %.2f" % logistic_running.score(test_set_running, test_label_categories))
print ("One vs Rest car accuracy is %.2f" % logistic_car.score(test_set_car, test_label_categories))
print ("One vs Rest stairs accuracy is %.2f" % logistic_stairs.score(test_set_stairs, test_label_categories))


======= Decision Tree Classifier =======
Accuracy for multiclass is 0.87
Predicted labels my MultiClass are: 	[0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 4, 4, 2, 3, 4, 3, 3, 3, 3, 4, 4, 4, 3, 4, 4]
Actual test labels should be: 		[0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4]
Where 0 is sitting, 1 is walking, 2 is running, 3 is car, 4 is stairs.

One vs Rest sitting accuracy is 1.00
One vs Rest walking accuracy is 1.00
One vs Rest running accuracy is 0.93
One vs Rest car accuracy is 0.93
One vs Rest stairs accuracy is 0.87

======= SVM Classifier =======
Accuracy for MultiClass is 0.83
Predicted labels my MultiClass are: 	[0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 4, 2, 3, 4, 4, 4, 3, 3, 4, 4, 4, 3, 4, 4]
Actual test labels should be: 		[0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4]
Where 0 is sitting, 1 is walking, 2 is running, 3 is car, 4 is stairs.

One vs Rest sitting accuracy is 1.00


- General Findings

When I had everything besides 'running', my multi-class accuracies for the 3 classifiers (with just the means of acceleration on the Z axis as the features) were around 93%. 
However, after I added the running data, my accuracies dropped to around 70%. I tried a few things until I found that a combination of the accelerations on all 3 axes works best (but still not as high as without running).

However, acceleration on the Y axis is what classified 'walking' with the highest accuracy in all classifiers. 

Acceleration on the X axis performed the worst from all classifiers. 

- Which features were most useful?
I combined acceleration on all 3 axes: abs(accelerationX) + accelerationY + accelerationZ to get a value for each line, and then took the mean of all 100 lines within one 10-second chunk. That gave me pretty high accuracies: 87% for decision trees, 83% for SVM, and 80% for Logistic Regression.

- Which learning algorithms were most effective?
The most effective one was decision trees. 

- Which activity was hardest to classify?
The activity that was hardest to classify was running. As mentioned before, when I introduced it to my dataset, my accuracies went down, so I had to change the features array. 
Sitting and Walking were always classified correctly. However, running, climbing stairs, and riding in a car were miss-classified.  I think that's because I went up and down the stairs really quickly, so it almost looks like I'm running. As for being in car, Providence's roads are very bumpy, but we also stopped at many traffic lights, so I can see how the movement in the car might have resembled running/climbing stairs. 
